src/Stratis.Bitcoin.Features.Wallet/WalletTransactionHandler.cs

```diff
@@ -55,20 +55,20 @@ public Transaction BuildTransaction(TransactionBuildContext context)
             if (context.Shuffle)
                 context.TransactionBuilder.Shuffle();
 
-            Transaction transaction = context.TransactionBuilder.BuildTransaction(false);
-            if (context.Sign)
-            {
-                ICoin[] coinsSpent = context.TransactionBuilder.FindSpentCoins(transaction);
-                this.AddSecrets(context, coinsSpent);
-                context.TransactionBuilder.SignTransactionInPlace(transaction);
-            }
-
             const int maxRetries = 3;
             int retryCount = 0;
 
             TransactionPolicyError[] errors = null;
             while (retryCount <= maxRetries)
             {
+                Transaction transaction = context.TransactionBuilder.BuildTransaction(false);
+                if (context.Sign)
+                {
+                    ICoin[] coinsSpent = context.TransactionBuilder.FindSpentCoins(transaction);
+                    this.AddSecrets(context, coinsSpent);
```

Comment:
We may have already added keys from the previous try, ensure that it works and doesn't rise exceptions.

Even if this works, this could be optimized to prevent the computation of keys we already computed in the previous try (because computing keys is one of the slowest part), but since this shouldn't happen so often you could leave as it is now, but add a TODO comment to flag this part as "needed to be improved"